File: 322-coin-change/coin-change.cpp
class Solution {
public:
    int solve(int i,int target,vector<int>& c,vector<vector<int>>&dp){
        if(target==0){
            return 0;
        }
        if(i==0){
            int cnt=0;
            while(c[0]<=target){
                target-=c[0];
                cnt++;
            }
            if(target==0)return cnt;
            return 1e9;
        }
        if(dp[i][target]!=-1)return dp[i][target];
        int nt=0+solve(i-1,target,c,dp);
        int t=1e9;
        if(target>=c[i])
        t=1+solve(i,target-c[i],c,dp);

        return dp[i][target]=min(nt,t);
    }
    int coinChange(vector<int>& c, int amount) {
        int n=c.size();
        vector<vector<int>>dp(n,vector<int>(amount+1,-1));
        int ans=solve(n-1,amount,c,dp);
        if(ans==1e9)return -1;
        return ans;
    }
};

File: 322-coin-change/coin-change.java
class Solution {
    public int solve(int i,int sum,int n,int[]coins,int[][]dp){
        if(i>=n){
            return (int)1e9;
        }
        if(i==n-1){
            if(sum%coins[i]==0){
                return dp[i][sum]=sum/coins[i];
            }
            else{
                return dp[i][sum]=(int)1e9;
            }
        }
        if(dp[i][sum]!=-1){
            return dp[i][sum];
        }
        int take=(int)1e9;
        if(coins[i]<=sum){
            take=1+solve(i,sum-coins[i],n,coins,dp);
        }
        int not_take = 0+ solve(i+1,sum,n,coins,dp);

        return dp[i][sum]=Math.min(take,not_take);

    }
    public int coinChange(int[] coins, int amount) {
        int n=coins.length;
        int [][]dp = new int[n][amount+1];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }
        int ans = solve(0,amount,n,coins,dp);
        if(ans>=(int)1e9){
            return -1;
        }
        else return ans;
    }
}